yt/frontends/amrvac/data_structures.py

```diff
@@ -250,16 +251,101 @@ def _set_code_unit_attributes(self):
+        # AMRVAC does not allow to specify any other normalisation beside those four.
+
+        # note: _override_code_units() has already been called, so self.units_override has been set.
+        # _override_code_units() sets supplied items in units_override as attributes.
+        # Handled items are length, time, mass, velocity, magnetic, temperature
+
+        # note: yt sets hydrogen mass equal to proton mass, amrvac doesn't.
+        mp_cgs = self.quan(1.672621898e-24, 'g')  # This value is taken from AstroPy
+        He_abundance = 0.1  # hardcoded parameter in AMRVAC
+
+        # get self.length_unit if overrides are supplied, otherwise use default
+        length_unit = getattr(self, 'length_unit', self.quan(1, 'cm'))
+
+        # 1. calculations for mass, density, numberdensity
+        if 'mass_unit' in self.units_override:
+            # in this case unit_mass is supplied (and has been set as attribute)
+            mass_unit = self.mass_unit
+            density_unit = mass_unit / length_unit**3
+            numberdensity_unit = density_unit / ((1.0 + 4.0 * He_abundance) * mp_cgs)
+        else:
+            # other case: numberdensity is supplied. Fall back to one (default) if no overrides supplied
+            numberdensity_override = self.units_override.get('numberdensity_unit', (1, 'cm**-3'))
+            numberdensity_unit = self.quan(*numberdensity_override)  # numberdensity is never set as attribute
+            density_unit = (1.0 + 4.0 * He_abundance) * mp_cgs * numberdensity_unit
+            mass_unit = density_unit * length_unit**3
+
+        # 2. calculations for velocity
+        if 'time_unit' in self.units_override:
+            # in this case time was supplied
+            velocity_unit = length_unit / self.time_unit
+        else:
+            # other case: velocity was supplied. Fall back to zero (default) if no overrides supplied
+            velocity_unit = getattr(self, 'velocity_unit', None)
+
+        # 3. calculations for pressure and temperature
+        if velocity_unit is None:
+            # velocity and time not given, see if temperature is given. Fall back to one (default) if not
+            temperature_unit = getattr(self, 'temperature_unit', self.quan(1, 'K'))
+            pressure_unit = ((2.0 + 3.0 * He_abundance) * numberdensity_unit * kb_cgs * temperature_unit).to('dyn*cm**-2')
+            velocity_unit = (np.sqrt(pressure_unit / density_unit)).to('cm*s**-1')
+        else:
+            # velocity is not zero if either time was given OR velocity was given
+            pressure_unit = (density_unit * velocity_unit ** 2).to('dyn*cm**-2')
+            temperature_unit = (pressure_unit / ((2.0 + 3.0 * He_abundance) * numberdensity_unit * kb_cgs)).to('K')
+
+        # 4. calculations for magnetic unit and time
+        time_unit = getattr(self, 'time_unit', length_unit / velocity_unit)  # if time given use it, else calculate
+        magnetic_unit = (np.sqrt(4 * np.pi * pressure_unit)).to('gauss')
```

Comment:
same as above.
I'm also wondering how that specific line would work out... since when is "gauss" equivalent to "dyne**1/2" ?

Reply:
Because the magnetic constant  `mu_0` is defined as 4pi in cgs?

Reply:
Wait no, this doesn't make sense. mu0 is not dimensionless, especially if you define it "_in  cgs_", which would be a meaningless statement if you were talking about a truly dimensionless quantity.